applications/plugins/SofaImplicitField/components/geometry/bottleField.cpp

```diff
@@ -0,0 +1,173 @@
+/******************************************************************************
+*                 SOFA, Simulation Open-Framework Architecture                *
+*                    (c) 2006 INRIA, USTL, UJF, CNRS, MGH                     *
+*                                                                             *
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#include <sofa/core/ObjectFactory.h>
+using sofa::core::RegisterObject ;
+
+#include "bottleField.h"
+
+namespace sofa
+{
+
+namespace component
+{
+
+namespace geometry
+{
+
+namespace _bottleField_
+{
+using sofa::defaulttype::Vec2;
+
+bottleField::bottleField()
+    : d_inside(initData(&d_inside, false, "inside", "If true the field is oriented inside (resp. outside) the bottle-shaped object. (default = false)"))
+    , d_radiusSphere(initData(&d_radiusSphere, 1.0, "radius", "Radius of Sphere emitting the field. (default = 1)"))
+    , d_centerSphere(initData(&d_centerSphere, Vec3d(0.0,0.0,0.0), "center", "Position of the Sphere Surface. (default=0 0 0)" ))
+    , d_shift(initData(&d_shift, 1.0, "shift", "How much the top ellipsoid is shifted from the bottom sphere. (default=1)" ))
+    , d_ellipsoidRadius(initData(&d_ellipsoidRadius, 1.0, "ellipsoidRadius", "Radius of the ellipsoid whose intersection with the sphere is taken off" ))
+    , d_excentricity(initData(&d_excentricity, 1.0, "excentricity", "excentricity of ellipsoid" ))
+{init();
+    }
+
+void bottleField::init()
+{
+    m_inside = d_inside.getValue();
```

Comment:
With this kind of code te internal variables are not updaded when variables changes. 
There is now a  mecahnism in Sofa to implement the correct behavior.
Drafted example:
``̀`cpp
 addUpdateCallback("myUpdateCallback", {&d_inside, &d_radius, }, [this](const core::DataTracker& t)
    {
        SOFA_UNUSED(t);
       this->init();
        return sofa::core::objectmodel::ComponentState::Valid;
    }, {});
```